src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -1620,22 +1627,119 @@ public boolean isRepaired()
+
+        try (FileDataInput in = dfile.createReader(position))
+        {
+            if (in.isEOF())
+                return null;
+
+            int version = descriptor.version.correspondingMessagingVersion();
+            DeserializationHelper helper = new DeserializationHelper(metadata.get(),
+                                                                     version,
+                                                                     DeserializationHelper.Flag.LOCAL,
+                                                                     columnFilter);
+
+            return UnfilteredSerializer.serializer.deserializeStaticRow(in, header, helper);
         }
+    }
 
-        return key;
+    /**
+     * Retrieves the clustering prefix of the unfiltered at the given position of the data file, as specified by
+     * {@link SSTableFlushObserver#nextUnfilteredCluster(Unfiltered, long)}.
+     *
+     * @param position the start position of the unfiltered in the data file
+     * @return the clustering prefix of the unfiltered at the specified position
+     */
+    public ClusteringPrefix clusteringAt(long position) throws IOException
+    {
+        try (FileDataInput in = dfile.createReader(position))
+        {
+            if (in.isEOF())
+                return null;
+
+            int version = descriptor.version.correspondingMessagingVersion();
+            int flags = in.readUnsignedByte();
+            boolean isRow = UnfilteredSerializer.kind(flags) == Unfiltered.Kind.ROW;
+
+            return isRow
+                   ? Clustering.serializer.deserialize(in, version, header.clusteringTypes())
+                   : ClusteringBoundOrBoundary.serializer.deserialize(in, version, header.clusteringTypes());
+        }
+    }
+
+    /**
+     * Retrieves the unfiltered at the given position of the data file, as specified by
+     * {@link SSTableFlushObserver#nextUnfilteredCluster(Unfiltered, long)}.
+     *
+     * @param position the start position of the unfiltered in the data file
+     * @param columnFilter the columns to fetch, {@code null} to select all the columns
+     * @return the unfiltered at the specified position
+     */
+    public Unfiltered unfilteredAt(long position, ColumnFilter columnFilter) throws IOException
+    {
+        try (FileDataInput in = dfile.createReader(position))
+        {
+            if (in.isEOF())
+                return null;
+
+            int version = descriptor.version.correspondingMessagingVersion();
+            DeserializationHelper helper = new DeserializationHelper(metadata.get(),
+                                                                     version,
+                                                                     DeserializationHelper.Flag.LOCAL,
+                                                                     columnFilter);
+            return UnfilteredSerializer.serializer.deserialize(in, header, helper, BTreeRow.sortedBuilder());
+        }
```

Comment:
The `partitionLevelDeletionAt()`/`staticRowAt()`/`clusteringAt()`/`unfilteredAt()` complex has neither users nor tests at this point, so it may make the most sense to leave them for a future patch that does?

Reply:
added some tests into `SSTableFlushObserverTest`